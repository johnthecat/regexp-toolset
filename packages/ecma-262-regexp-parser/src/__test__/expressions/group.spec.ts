import { expect, describe, it } from 'vitest';
import { parseRegexp } from '../../index.js';

/**
 * @see parseGroup
 */
describe('Group', () => {
  describe('Common', () => {
    it('should throw, if group is not closed', () => {
      expect(() => parseRegexp('/(ab/')).toThrowErrorMatchingInlineSnapshot(`
        "
         ❯ /(ab/
            ════
         Incomplete group structure"
      `);
    });

    it('should throw, there is unmatched parenthesis', () => {
      expect(() => parseRegexp('/(ab))/')).toThrowErrorMatchingInlineSnapshot(`
        "
         ❯ /(ab))/
                ↑
         Unmatched parenthesis"
      `);
    });

    it('should correctly parse nested groups', () => {
      expect(parseRegexp('/(a(f)b)/')).toMatchSnapshot();
    });
  });

  describe.each([
    { name: 'Capturing Group', prefix: '' },
    { name: 'Named Capturing Group', prefix: '?<my_name>' },
    { name: 'Non Capturing Group', prefix: '?:' },
    { name: 'Positive lookahead', prefix: '?=' },
    { name: 'Positive lookbehind', prefix: '?<=' },
    { name: 'Negative lookahead', prefix: '?!' },
    { name: 'Negative lookbehind', prefix: '?<!' },
  ])('$name', ({ prefix }) => {
    it('should parse simple', () => {
      expect(parseRegexp(new RegExp(`(${prefix}a)`))).toMatchSnapshot();
    });
  });
});
