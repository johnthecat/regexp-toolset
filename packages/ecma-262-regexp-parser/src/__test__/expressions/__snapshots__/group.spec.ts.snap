// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Group > 'Capturing Group' > should parse simple 1`] = `
{
  "body": {
    "body": {
      "charCode": 97,
      "end": 2,
      "kind": "Char",
      "start": 2,
      "type": "Simple",
      "value": "a",
    },
    "end": 3,
    "index": 1,
    "kind": "Group",
    "specifier": null,
    "start": 1,
    "type": "capturing",
  },
  "end": 4,
  "flags": "",
  "kind": "Regexp",
  "start": 0,
}
`;

exports[`Group > 'Named Capturing Group' > should parse simple 1`] = `
{
  "body": {
    "body": {
      "charCode": 97,
      "end": 12,
      "kind": "Char",
      "start": 12,
      "type": "Simple",
      "value": "a",
    },
    "end": 13,
    "index": 1,
    "kind": "Group",
    "specifier": {
      "end": 11,
      "kind": "GroupName",
      "name": "my_name",
      "start": 3,
    },
    "start": 1,
    "type": "capturing",
  },
  "end": 14,
  "flags": "",
  "kind": "Regexp",
  "start": 0,
}
`;

exports[`Group > 'Negative lookahead' > should parse simple 1`] = `
{
  "body": {
    "body": {
      "charCode": 97,
      "end": 4,
      "kind": "Char",
      "start": 4,
      "type": "Simple",
      "value": "a",
    },
    "end": 5,
    "kind": "Group",
    "start": 1,
    "type": "negativeLookahead",
  },
  "end": 6,
  "flags": "",
  "kind": "Regexp",
  "start": 0,
}
`;

exports[`Group > 'Negative lookbehind' > should parse simple 1`] = `
{
  "body": {
    "body": {
      "charCode": 97,
      "end": 5,
      "kind": "Char",
      "start": 5,
      "type": "Simple",
      "value": "a",
    },
    "end": 6,
    "kind": "Group",
    "start": 1,
    "type": "negativeLookbehind",
  },
  "end": 7,
  "flags": "",
  "kind": "Regexp",
  "start": 0,
}
`;

exports[`Group > 'Non Capturing Group' > should parse simple 1`] = `
{
  "body": {
    "body": {
      "charCode": 97,
      "end": 4,
      "kind": "Char",
      "start": 4,
      "type": "Simple",
      "value": "a",
    },
    "end": 5,
    "kind": "Group",
    "start": 1,
    "type": "nonCapturing",
  },
  "end": 6,
  "flags": "",
  "kind": "Regexp",
  "start": 0,
}
`;

exports[`Group > 'Positive lookahead' > should parse simple 1`] = `
{
  "body": {
    "body": {
      "charCode": 97,
      "end": 4,
      "kind": "Char",
      "start": 4,
      "type": "Simple",
      "value": "a",
    },
    "end": 5,
    "kind": "Group",
    "start": 1,
    "type": "positiveLookahead",
  },
  "end": 6,
  "flags": "",
  "kind": "Regexp",
  "start": 0,
}
`;

exports[`Group > 'Positive lookbehind' > should parse simple 1`] = `
{
  "body": {
    "body": {
      "charCode": 97,
      "end": 5,
      "kind": "Char",
      "start": 5,
      "type": "Simple",
      "value": "a",
    },
    "end": 6,
    "kind": "Group",
    "start": 1,
    "type": "positiveLookbehind",
  },
  "end": 7,
  "flags": "",
  "kind": "Regexp",
  "start": 0,
}
`;

exports[`Group > Common > should correctly parse nested groups 1`] = `
{
  "body": {
    "body": {
      "end": 6,
      "expressions": [
        {
          "charCode": 97,
          "end": 2,
          "kind": "Char",
          "start": 2,
          "type": "Simple",
          "value": "a",
        },
        {
          "body": {
            "charCode": 102,
            "end": 4,
            "kind": "Char",
            "start": 4,
            "type": "Simple",
            "value": "f",
          },
          "end": 5,
          "index": 2,
          "kind": "Group",
          "specifier": null,
          "start": 3,
          "type": "capturing",
        },
        {
          "charCode": 98,
          "end": 6,
          "kind": "Char",
          "start": 6,
          "type": "Simple",
          "value": "b",
        },
      ],
      "kind": "Alternative",
      "start": 2,
    },
    "end": 7,
    "index": 1,
    "kind": "Group",
    "specifier": null,
    "start": 1,
    "type": "capturing",
  },
  "end": 8,
  "flags": "",
  "kind": "Regexp",
  "start": 0,
}
`;
